File: crates/ontologos-dotnet/csharp/src/Ontologos/Reasoner.cs
namespace Ontologos;

/// <summary>OWL reasoner handle.</summary>
public sealed class Reasoner : IDisposable
{
    private long _handle;
    private bool _disposed;

    static Reasoner()
    {
        NativeLoader.EnsureLoaded();
    }

    private Reasoner(long handle)
    {
        _handle = NativeInterop.CheckHandle(handle, "reasoner");
    }

    public Reasoner(Ontology ontology, string? profile, bool incremental = false, long? budgetSecs = null)
        : this(
            NativeMethods.ReasonerNewNative(
                ontology.NativeHandle(),
                profile,
                incremental ? 1 : 0,
                NativeInterop.ToNativeOptional(budgetSecs)))
    {
    }

    public static Reasoner FromPath(
        string path,
        string? profile,
        bool incremental = false,
        long? budgetSecs = null,
        bool lenient = false) =>
        new(
            NativeMethods.ReasonerFromPathNative(
                path,
                profile,
                incremental ? 1 : 0,
                NativeInterop.ToNativeOptional(budgetSecs),
                lenient ? 1 : 0));

    public static Reasoner LoadIn(
        string baseDir,
        string path,
        string? profile,
        bool incremental = false,
        long? budgetSecs = null,
        bool lenient = false) =>
        new(
            NativeMethods.ReasonerLoadInNative(
                baseDir,
                path,
                profile,
                incremental ? 1 : 0,
                NativeInterop.ToNativeOptional(budgetSecs),
                lenient ? 1 : 0));

    public string Classify()
    {
        EnsureOpen();
        return NativeInterop.TakeString(NativeMethods.ReasonerClassifyNative(_handle));
    }

    public string Explain()
    {
        EnsureOpen();
        return NativeInterop.TakeString(NativeMethods.ReasonerExplainNative(_handle));
    }

    public string ParseMeta()
    {
        EnsureOpen();
        return NativeInterop.TakeString(NativeMethods.ReasonerParseMetaNative(_handle));
    }

    public string? Taxonomy()
    {
        EnsureOpen();
        return NativeInterop.TakeString(NativeMethods.ReasonerTaxonomyNative(_handle), allowNull: true);
    }

    public string CheckConsistency()
    {
        EnsureOpen();
        return NativeInterop.TakeString(NativeMethods.ReasonerCheckConsistencyNative(_handle));
    }

    public bool IsConsistent()
    {
        EnsureOpen();
        return NativeMethods.ReasonerIsConsistentNative(_handle) != 0;
    }

    public bool IsEntailed(EntailmentCheck check)
    {
        ArgumentNullException.ThrowIfNull(check);
        EnsureOpen();
        return NativeMethods.ReasonerIsEntailedNative(
            _handle,
            check.Sub,
            check.Sup,
            check.Individual,
            check.ClassIri,
            check.Subject,
            check.Property,
            check.Object) != 0;
    }

    public string Query(string query)
    {
        EnsureOpen();
        return NativeInterop.TakeString(NativeMethods.ReasonerQueryNative(_handle, query));
    }

    public Reasoner AddSubclassOf(string subclass, string superclass)
    {
        EnsureOpen();
        _handle = NativeMethods.ReasonerAddSubclassOfNative(_handle, subclass, superclass);
        return this;
    }

    public Reasoner RemoveSubclassOf(string subclass, string superclass)
    {
        EnsureOpen();
        _handle = NativeMethods.ReasonerRemoveSubclassOfNative(_handle, subclass, superclass);
        return this;
    }

    public Reasoner AddAxiomJson(string axiomJson)
    {
        EnsureOpen();
        _handle = NativeMethods.ReasonerAddAxiomJsonNative(_handle, axiomJson);
        return this;
    }

    public void Dispose()
    {
        if (!_disposed && _handle != 0)
        {
            NativeMethods.ReasonerCloseNative(_handle);
            _handle = 0;
            _disposed = true;
        }
    }

    private void EnsureOpen()
    {
        if (_disposed || _handle == 0)
        {
            throw new ObjectDisposedException(nameof(Reasoner));
        }
    }
}

File: crates/ontologos-java/java/src/main/java/dev/ontologos/Reasoner.java
package dev.ontologos;

import java.util.Objects;

/** OWL reasoner handle. */
public final class Reasoner implements AutoCloseable {
    static {
        NativeLoader.load();
    }

    private long handle;
    private boolean closed;

    private Reasoner(long handle) {
        if (handle == 0L) {
            throw new OntologosException("failed to create reasoner");
        }
        this.handle = handle;
    }

    public Reasoner(Ontology ontology, String profile) {
        this(ontology, profile, false, null);
    }

    public Reasoner(Ontology ontology, String profile, boolean incremental, Long budgetSecs) {
        this(
                nativeNew(
                        Objects.requireNonNull(ontology).nativeHandle(),
                        profile,
                        incremental,
                        toNativeOptional(budgetSecs)));
    }

    public static Reasoner fromPath(String path, String profile) {
        return fromPath(path, profile, false, null, false);
    }

    public static Reasoner fromPath(
            String path, String profile, boolean incremental, Long budgetSecs, boolean lenient) {
        return new Reasoner(
                nativeFromPath(
                        Objects.requireNonNull(path),
                        profile,
                        incremental,
                        toNativeOptional(budgetSecs),
                        lenient));
    }

    public static Reasoner loadIn(String base, String path, String profile) {
        return loadIn(base, path, profile, false, null, false);
    }

    public static Reasoner loadIn(
            String base,
            String path,
            String profile,
            boolean incremental,
            Long budgetSecs,
            boolean lenient) {
        return new Reasoner(
                nativeLoadIn(
                        Objects.requireNonNull(base),
                        Objects.requireNonNull(path),
                        profile,
                        incremental,
                        toNativeOptional(budgetSecs),
                        lenient));
    }

    /** Returns classify/materialize report JSON. */
    public String classify() {
        ensureOpen();
        return nativeClassify(handle);
    }

    /** Returns explain result JSON. */
    public String explain() {
        ensureOpen();
        return nativeExplain(handle);
    }

    /** Returns parse metadata JSON (empty object when absent). */
    public String parseMeta() {
        ensureOpen();
        return nativeParseMeta(handle);
    }

    /** Returns taxonomy JSON after EL/DL classify, or {@code null}. */
    public String taxonomy() {
        ensureOpen();
        return nativeTaxonomy(handle);
    }

    /** Returns consistency check JSON. */
    public String checkConsistency() {
        ensureOpen();
        return nativeCheckConsistency(handle);
    }

    public boolean isConsistent() {
        ensureOpen();
        return nativeIsConsistent(handle);
    }

    public boolean isEntailed(EntailmentCheck check) {
        ensureOpen();
        Objects.requireNonNull(check);
        return nativeIsEntailed(
                handle,
                check.sub(),
                check.sup(),
                check.individual(),
                check.classIri(),
                check.subject(),
                check.property(),
                check.object());
    }

    /** Returns query bindings JSON array. */
    public String query(String query) {
        ensureOpen();
        return nativeQuery(handle, Objects.requireNonNull(query));
    }

    public Reasoner addSubclassOf(String subclass, String superclass) {
        ensureOpen();
        handle =
                nativeAddSubclassOf(
                        handle,
                        Objects.requireNonNull(subclass),
                        Objects.requireNonNull(superclass));
        return this;
    }

    public Reasoner removeSubclassOf(String subclass, String superclass) {
        ensureOpen();
        handle =
                nativeRemoveSubclassOf(
                        handle,
                        Objects.requireNonNull(subclass),
                        Objects.requireNonNull(superclass));
        return this;
    }

    public Reasoner addAxiomJson(String axiomJson) {
        ensureOpen();
        handle = nativeAddAxiomJson(handle, Objects.requireNonNull(axiomJson));
        return this;
    }

    @Override
    public void close() {
        if (!closed && handle != 0L) {
            nativeClose(handle);
            handle = 0L;
            closed = true;
        }
    }

    private void ensureOpen() {
        if (closed || handle == 0L) {
            throw new IllegalStateException("reasoner handle is closed");
        }
    }

    private static long toNativeOptional(Long value) {
        return value == null ? -1L : value;
    }

    private static native long nativeNew(
            long ontologyHandle, String profile, boolean incremental, long budgetSecs);

    private static native long nativeFromPath(
            String path, String profile, boolean incremental, long budgetSecs, boolean lenient);

    private static native long nativeLoadIn(
            String base,
            String path,
            String profile,
            boolean incremental,
            long budgetSecs,
            boolean lenient);

    private static native String nativeParseMeta(long handle);

    private static native String nativeTaxonomy(long handle);

    private static native String nativeClassify(long handle);

    private static native String nativeExplain(long handle);

    private static native String nativeCheckConsistency(long handle);

    private static native boolean nativeIsConsistent(long handle);

    private static native boolean nativeIsEntailed(
            long handle,
            String sub,
            String sup,
            String individual,
            String classIri,
            String subject,
            String property,
            String object);

    private static native String nativeQuery(long handle, String query);

    private static native long nativeAddSubclassOf(long handle, String subclass, String superclass);

    private static native long nativeRemoveSubclassOf(long handle, String subclass, String superclass);

    private static native long nativeAddAxiomJson(long handle, String axiomJson);

    private static native void nativeClose(long handle);
}
